docs/source/typing_anti_pitch.rst

```diff
@@ -0,0 +1,80 @@
+in Python::
+
+* You simply don't want to. Python is a tool that is meant to serve you. Python is a big tent,
+  multi-paradigm language that generally allows you to do things in the way that best suits your
+  needs, as best determined by you.
+
+* The cost-benefit ratio isn't good enough. Pleasing static type checkers requires a non-zero amount
+  of busy work. If this isn't worth the extra confidence you get, you shouldn't add static type
+  checking.
+
+* Your codebase fits in your developers' heads. Opinions vary, but people tend to agree that at
+  some number of developers and lines of code, static type checking confers significantly more
+  benefit. You don't feel like you're there yet.
+
+* If you maintain high test coverage, that might provide sufficient quality assurance for your
+  needs (acknowledging that static type checking and tests enforce different things; static type
+  checking usually cannot validate logic, tests can often not prove invariants of your code to
+  hold).
+
+* Your codebase is old, large and has been working fine without static type checking for years.
+  While Python's type system is designed to
+  `allow gradual adoption of static type checking <https://mypy.readthedocs.io/en/stable/existing_code.html>`_,
+  the total cost of adding type annotations to a large extant codebase can be prohibitive.
+
+* Your library does enough dynamic things that type checking would be unlikely to help your users.
+  Either you would have to spend a lot of effort redesigning your library in ways that static type
+  checkers could better understand, or you'd have to invest a lot of effort into figuring out clever
+  type annotations to twist the arms of type checkers.
+
+* Your codebase has suffered at the hands of `Hyrum's Law <https://www.hyrumslaw.com/>`_
+  and all possible observable behaviour is depended on. In order to avoid false positives for your
+  users, all your types end up being either a) complicated ``Protocol``\s that are hard to maintain,
+  or b) ``Any`` in which case there's not much point. (On the other hand, static type checking could
+  be a good solution for communicating to users what behaviour they should be allowed to rely on)
+
+* You're not opposed to type checking in theory, but you dislike Python type checkers in practice.
+  Maybe they don't understand enough of the idioms you use, maybe you'd like them to infer more
+  instead of relying on explicit annotations, maybe they're too slow, maybe they don't integrate
+  well with your editor, maybe they're too hard to configure. Whatever the reason -- it just doesn't
+  work for your project.
+
+* Type checking in Python isn't actually strict enough, powerful enough or expressive enough for
+  you. Python type checkers end up making various decisions out of pragmatism, or due to limited
+  resources, and these decisions might not be the ones for you.
+
+Advice for maintainers of untyped libraries
+*******************************************
```

Comment:
As first point I would mention the possibility to distribute a stub file directly with the library. This is the best solution for users, while not using type checking for the library itself. The other options are more suited if a maintainer doesn't want that burden either.

Reply:
No, I tried that with `editables` and apparently (I don't know whether this is still true) mypy doesn't typecheck the library code against the stubs. Which means there's literally zero benefit to the library maintainer for shipping a stubs file.

Also, it still puts the burden on the library maintainer and the point of this document is to *reduce* the burden on maintainers who don't want to add typing. The message should be "if all you (as a user) want is to have type annotations, you can write them yourself so you don't need the library maintainer to do *anything*".

Reply:
The idea is that the library code is not checked against the stubs. But it is shipped with the library, which has clear benefits to the library's users.

And of course this is more burden for the library maintainer than just handing off writing the stubs to other people, but it is a valid option, and the best one if a maintainer doesn't want to add type checking to the library, but is willing to maintain the stubs.

Reply:
I added a separate mention of this and made sure to clarify the limits. This could be useful for maintainers whose main objection is readability